platform/macos/joypad_macos.mm

```diff
@@ -337,257 +309,100 @@ - (void)addMacOSJoypad:(GCController *)controller {
 
 	// Assign our player index.
 	if (controller.playerIndex == GCControllerPlayerIndexUnset) {
-		controller.playerIndex = [self getFreePlayerIndex];
+		controller.playerIndex = get_free_player_index();
 	}
 
 	// Tell Godot about our new controller.
-	Input::get_singleton()->joy_connection_changed(joy_id, true, String::utf8([controller.vendorName UTF8String]));
-
-	Joypad *joypad = [[Joypad alloc] init:controller];
-
-	// Add it to our dictionary, this will retain our controllers.
-	[self.connectedJoypads setObject:joypad forKey:[NSNumber numberWithInt:joy_id]];
+	Input::get_singleton()->joy_connection_changed(joy_id, true, String::utf8(controller.vendorName.UTF8String));
 
-	// Set our input handler.
-	[self setControllerInputHandler:controller];
+	// Assign our player index.
+	joypads.insert(joy_id, memnew(GameController(joy_id, controller)));
+	controller_to_joy_id.insert(controller, joy_id);
 }
 
-- (void)controllerWasConnected:(NSNotification *)notification {
-	// Get our controller.
-	GCController *controller = (GCController *)notification.object;
-
-	if (!controller) {
-		print_verbose("Couldn't retrieve new controller.");
+void JoypadMacOS::remove_joypad(GCController *controller) {
+	if (!controller_to_joy_id.has(controller)) {
 		return;
 	}
 
-	if ([[self getAllKeysForController:controller] count] > 0) {
-		print_verbose("Controller is already registered.");
-	} else if (!self.isProcessing) {
-		Joypad *joypad = [[Joypad alloc] init:controller];
-		[self.joypadsQueue addObject:joypad];
```

Comment:
This was a bug, as it was adding a `Joypad *` to `joypadsQueue`, which was expecting a `GCController *`, per:

https://github.com/godotengine/godot/blob/e25f3c0d38d457b15a63720240736f564ce0501b/platform/macos/joypad_macos.mm#L255-L257